File: solutions/2056. Number of Valid Move Combinations On Chessboard/2056.cpp
class Solution {
 public:
  int countCombinations(vector<string>& pieces,
                        vector<vector<int>>& positions) {
    const int n = pieces.size();
    unordered_set<long long> hashedBoards;
    // Stores all possible move combinations for `pieces`. Each element is a
    // vector of moves, one for each piece in the input order. e.g., if pieces =
    // ["rook", "bishop"], one element might be [[1,0], [1,1]], representing a
    // rook moving right and a bishop moving diagonally up-right.
    vector<vector<pair<int, int>>> pieceMovesList;

    getPieceMovesList(pieces, 0, {}, pieceMovesList);

    for (const vector<pair<int, int>>& pieceMoves : pieceMovesList)
      dfs(positions, n, pieceMoves, (1 << n) - 1, hashedBoards);

    return hashedBoards.size();
  }

 private:
  const unordered_map<string, vector<pair<int, int>>> kPieceToMoves{
      {"rook", {{1, 0}, {-1, 0}, {0, 1}, {0, -1}}},
      {"bishop", {{1, 1}, {1, -1}, {-1, 1}, {-1, -1}}},
      {"queen",
       {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}}}};

  // Generates all possible combinations of moves.
  void getPieceMovesList(const vector<string>& pieces, int i,
                         vector<pair<int, int>>&& path,
                         vector<vector<pair<int, int>>>& pieceMovesList) {
    if (i == pieces.size()) {
      pieceMovesList.push_back(path);
      return;
    }
    for (const pair<int, int>& move : kPieceToMoves.at(pieces[i])) {
      path.push_back(move);
      getPieceMovesList(pieces, i + 1, std::move(path), pieceMovesList);
      path.pop_back();
    }
  }

  // Performs a depth-first search to explore all possible board states.
  void dfs(const vector<vector<int>>& board, int n,
           const vector<pair<int, int>>& pieceMoves, int activeMask,
           unordered_set<long long>& hashedBoards) {
    if (activeMask == 0)
      return;
    hashedBoards.insert(getHash(board));
    for (int nextActiveMask = 1; nextActiveMask < 1 << n; ++nextActiveMask) {
      if ((activeMask & nextActiveMask) != nextActiveMask)
        continue;

      // Copy the board.
      vector<vector<int>> nextBoard = board;

      // Move the pieces that are active in this turn.
      for (int i = 0; i < n; ++i)
        if (nextActiveMask >> i & 1) {
          nextBoard[i][0] += pieceMoves[i].first;
          nextBoard[i][1] += pieceMoves[i].second;
        }

      // No two or more pieces occupy the same square.
      if (getUniqueSize(nextBoard) < n)
        continue;

      // Every piece needs to be in the boundary.
      if (ranges::all_of(nextBoard, [](const vector<int>& pos) {
        return 1 <= pos[0] && pos[0] <= 8 && 1 <= pos[1] && pos[1] <= 8;
      }))
        dfs(nextBoard, n, pieceMoves, nextActiveMask, hashedBoards);
    }
  }

  long long getHash(const vector<vector<int>>& board) {
    long long hash = 0;
    for (const vector<int>& pos : board)
      hash = (hash * 64) + ((pos[0] - 1) << 3) + (pos[1] - 1);
    return hash;
  }

  int getUniqueSize(const vector<vector<int>>& board) {
    unordered_set<int> unique;
    for (const vector<int>& pos : board)
      unique.insert(pos[0] * 8 + pos[1]);
    return unique.size();
  }
};

File: solutions/2056. Number of Valid Move Combinations On Chessboard/2056.py
class Solution:
  def countCombinations(
      self,
      pieces: list[str],
      positions: list[list[int]],
  ) -> int:
    n = len(pieces)
    moves = {"rook": [(1, 0), (-1, 0), (0, 1), (0, -1)],
             "bishop": [(1, 1), (1, -1), (-1, 1), (-1, -1)],
             "queen": [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]}
    hashedBoards = set()

    def getHash(board: list[list[int]]) -> Tuple:
      return tuple([tuple(pos) for pos in board])

    def dfs(
        board: list[list[int]],
        pieceMoves: list[tuple[int, int]],
        activeMask: int,
    ) -> None:
      """Performs a depth-first search to explore all possible board states."""
      if activeMask == 0:
        return
      hashedBoards.add(getHash(board))

      for nextActiveMask in range(1, 1 << n):
        if activeMask & nextActiveMask != nextActiveMask:
          continue

        # Copy the board.
        nextBoard = [pos.copy() for pos in board]

        # Move the pieces that are active in this turn.
        for i in range(n):
          if nextActiveMask >> i & 1:
            nextBoard[i][0] += pieceMoves[i][0]
            nextBoard[i][1] += pieceMoves[i][1]

        # No two or more pieces occupy the same square.
        if len(set(getHash(nextBoard))) < n:
          continue

        # Every piece needs to be in the boundary.
        if all(1 <= x <= 8 and 1 <= y <= 8 for x, y in nextBoard):
          dfs(nextBoard, pieceMoves, nextActiveMask)

    for pieceMoves in itertools.product(*(moves[piece] for piece in pieces)):
      dfs(positions, pieceMoves, (1 << n) - 1)

    return len(hashedBoards)
